### PR TITLE
Azure POC

### PR DESCRIPTION
**Pull request by JunqiZhang0**
Added initial Azure support which allow creation and destruction

### CODE REVIEW

File: linchpin/.vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "python.pythonPath": "/bin/python"
```
Comment (abraverm) [inline linchpin/.vscode/settings.json]: remove this file

File: linchpin/provision/roles/azure/handlers/main.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+---
```
Comment (abraverm) [inline linchpin/provision/roles/azure/handlers/main.yml]: Is this file required?

File: linchpin/provision/roles/azure/tasks/teardown_res_defs.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+---
+- name: "teardown of Azure VM resource definitions of current group"
+  include: teardown_azure_vm.yml res_def={{ res_item.0 }} tp_out={{ res_item.1 }} res_grp_name={{ res_item.2 }}
+  when: res_item.0['role'] == "aws_ec2"
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/teardown_res_defs.yml]: typo

File: linchpin/provision/roles/azure/tasks/teardown_resource_group.yml [hunk 4]
```diff
@@ -0,0 +1,61 @@
+---
+- name: "Set cred profile"
+  set_fact:
+    cred_profile: "{{ res_grp['credentials']['profile'] | default('default') }}"
+  when: res_grp['credentials'] is defined
+
+- name: "Get creds from auth driver"
+  auth_driver:
+    filename: "{{ res_grp['credentials']['filename'] }}"
+    cred_type: "aws"
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/teardown_resource_group.yml]: typo

File: linchpin/provision/roles/azure/vars/ex_aws_creds.yml [hunk 5]
```diff
@@ -0,0 +1,2 @@
+aws_access_key_id: XXXXXXXXXXXXXXXXXXXXXX
```
Comment (abraverm) [inline linchpin/provision/roles/azure/vars/ex_aws_creds.yml]: aws??

File: linchpin/provision/roles/azure/tasks/provision_res_defs.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    asyn ac_outputs_azure_vm: []
+  when: _async
+
+- name: "Get the resource name"
+  set_fact:
+    rsrc_name: "{{ res_def['res_name'] | default(res_def['vm_name']) }}"
+
+- name: "Create name using uhash value"
+  set_fact:
+    aws_resource_name: "{{ rsrc_name + '-' + uhash + '-' }}"
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/provision_res_defs.yml]: typo

File: linchpin/provision/roles/azure/tasks/provision_azure_vm.yml [hunk 7]
```diff
@@ -0,0 +1,77 @@
+---
+- name: "Set default res_grp_vars when not defined"
+  set_fact:
+    res_grp_vars:
+      - name: "{{ res_def['res_name'] | default(res_def['vm_name']) }}"
+        resource_group_name: "{{ res_grp_name  }}"
+
+# - name: Update Instance Tags
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/provision_azure_vm.yml]: needed?

Comment (JunqiZhang0) [reply linchpin/provision/roles/azure/tasks/provision_azure_vm.yml]: yes

**Comment by JunqiZhang0** [comment]
Adding layout support,  virtual_network support, network_interface support
force user specify the network_name so that nothing is messed up
add location

File: linchpin/provision/roles/azure/files/schema.json [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+                            "type": "string",
+                            "required": true
+                        },
+                        "image": {
+                            "type": "string",
+                            "required": true
+                        },
+                        "vm_username": {
+                            "type": "string",
+                            "required": true
```
Comment (14rcole) [inline linchpin/provision/roles/azure/files/schema.json]: Are all of these fields really required or are there some for which we can set reasonable defaults?  e.g. {{ ansible_ssh_user }} for vm_username

Comment (JunqiZhang0) [reply linchpin/provision/roles/azure/files/schema.json]: yes, will do

**Comment by JunqiZhang0** [comment]
> Can you add documentation?

sure

File: linchpin/provision/roles/azure/tasks/provision_azure_vm.yml (hunk 7 above)
Comment (abraverm) [reply linchpin/provision/roles/azure/tasks/provision_azure_vm.yml]: please elaborate

File: docs/source/azure.rst [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+| role             | true       | string        | N/A               |                 |
++------------------+------------+---------------+-------------------+-----------------+
+| vm_name          | true       | string        | name              |It can't include |
+|                  |            |               |                   | _ and other     |
+|                  |            |               |                   |special char     |
++------------------+------------+---------------+-------------------+-----------------+
+| image            | true       | string        | image             |                 |
++------------------+------------+---------------+-------------------+-----------------+
+
+EC2 Inventory Generation
```
Comment (abraverm) [inline docs/source/azure.rst]: I dodn't know Azure has EC2 :)

File: docs/source/examples/workspaces/azure/PinFile [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+            role: azure_vm
+            resource_group: "ccit"
+            vm_size: "Standard_DS1_v2"
+            image: "rhcostestimage"
+#vm name can't include _
+            vm_username: {{"VM_NAME"|default("linchpinvm")}}
+            vm_password: {{"vm_password"|default( Asdfghjkl;']) }}
+        credentials:
+          filename: azure.key
+          profile: key
```
Comment (abraverm) [inline docs/source/examples/workspaces/azure/PinFile]: Add layout?

File: linchpin/provision/roles/azure/tasks/provision_azure_vm.yml (hunk 7 above)
Comment (JunqiZhang0) [reply linchpin/provision/roles/azure/tasks/provision_azure_vm.yml]: deleted

**Comment by JunqiZhang0** [comment]
@samvarankashyap @14rcole need to add the credential to jenkins for integration test, otherwise the test will fail

**Comment by JunqiZhang0** [comment]
**linchpin log** 
http://pastebin.test.redhat.com/793981

**inventory:**
[example]
40.122.112.244 hostname=40.122.112.244

[all]
40.122.112.244 hostname=40.122.112.244

**/resource/azure_vm.output**
http://pastebin.test.redhat.com/793986

**Comment by JunqiZhang0** [comment]
[merge]

**Comment by abraverm** [comment]
@JunqiZhang0, Please squash the commits

**Comment by JunqiZhang0** [comment]
[merge]

**Comment by samvarankashyap** [comment]
[merge]

**Comment by samvarankashyap** [comment]
[merge]

**Comment by samvarankashyap** [comment]
[merge]

**Pull request merged**